Sources/Body.swift

```diff
@@ -0,0 +1,251 @@
+     */
+    public mutating func becomeReceiver() throws -> ReceivingStream {
+        switch self {
+        case .receiver(let stream):
+            return stream
+        case .buffer(let data):
+            let stream = Drain(for: data)
+            self = .receiver(stream)
+            return stream
+        case .sender(let sender):
+            let stream = Drain()
+            try sender(stream)
+            self = .receiver(stream)
+            return stream
+        default:
+            throw BodyError.inconvertibleType
+        }
+    }
+
+    ///Returns true if body is case `receiver`
+    public var isReceiver: Bool {
+        switch self {
+        case .receiver: return true
+        default: return false
+        }
+    }
+
+    /**
+     Converts the body's contents into a closure
+     that accepts a `SendingStream`.
+     */
+    public mutating func becomeSender(timingOut deadline: Double = .never) -> ((SendingStream) throws -> Void) {
+        switch self {
+        case .buffer(let data):
+            let closure: ((SendingStream) throws -> Void) = { sender in
+                try sender.send(data, timingOut: deadline)
+            }
+            self = .sender(closure)
+            return closure
+        case .receiver(let receiver):
+            let closure: ((SendingStream) throws -> Void) = { sender in
+                let data = Drain(for: receiver, timingOut: deadline).data
+                try sender.send(data, timingOut: deadline)
+            }
+            self = .sender(closure)
+            return closure
+        case .sender(let sender):
+            return sender
+        default:
+            let closure: ((SendingStream) throws -> Void) = { _ in
+                throw BodyError.inconvertibleType
+            }
+            return closure
+        }
+    }
+
+    ///Returns true if body is case `sender`
+    public var isSender: Bool {
+        switch self {
+        case .sender: return true
+        default: return false
+        }
+    }
+}
+
+extension Body {
+    /**
+     Converts the body's contents into a `Data` buffer asynchronously.
+
+     If the body is a receiver, sender, asyncReceiver or asyncSender type,
+     it will be drained.
+     */
+    public func asyncBecomeBuffer(timingOut deadline: Double = .never, completion: ((Void) throws -> (Body, Data)) -> Void) {
+        switch self {
+        case .asyncReceiver(let stream):
+            _ = AsyncDrain(for: stream, timingOut: deadline) { closure in
+                completion {
+                    let drain = try closure ()
+                    return (.buffer(drain.data), drain.data)
+                }
+            }
+
+        case .asyncSender(let sender):
+            let drain = AsyncDrain()
+            sender(drain) { closure in
+                completion {
+                    try closure()
+                    return (.buffer(drain.data), drain.data)
+                }
+            }
+
+        default:
```

Comment:
This is awesome!

> If you took care of my future PR for about it(that talked on Slack), sorry for my meddling.
Sorry. I didn't understand this part.

I added you as a contributor to Zewo organization. If you want you can pull this branch (`quark`) and commit to it directly.